### DISCUSSION TITLE
Task 2, 03.01 - MODEL GENERATION

### DISCUSSION THREAD

**Discussion by valenrami5** [rramosp/ai4eng.v1#110] [03 - GENERACION DE MODELOS]
Buenas noches, tengo un problema, creo yo que con el autograder del task 2 del lab 03.01. 
Desarrolle el codigo como se puede observar en la primera foto y corre correctamente, cuando chequeo la solucion con el codigo dado, me da un accuracy de 0.9 como se debia esperar, por lo que creo que el codigo esta correcto.
![image](https://user-images.githubusercontent.com/98786597/151893666-ae2d7c96-f798-4c6f-8799-b29251950bfa.png)
Sin embargo, cuando intento pasar mi codigo por el autograder, me aparece que el nombre product no esta definido.
![image](https://user-images.githubusercontent.com/98786597/151893785-fe356784-fd4d-4d38-b2ca-170966362245.png)
Me gustaria saber que puedo hacer en este caso?

**Comment by Felipe-RA** [comment]
Saludos valenrami5

Como indica el mensaje de error, no se encuentra definido el import de `product` a la hora de evaluar en el grader tu función.

<ins>La explicación corta de tu problema es que debes importar las librerias que necesites **dentro de la función en la que las utilizas** (esto se explica en los vídeos)</ins>

Ahora Vamos a entender por qué esto pasa con una explicación un poco más larga:

Primero, recordarte la sintaxis de lo que se espera que sometas a evaluación:

```
def fit(X,y):
    def predict(X, t):
        return ...

    return ...

```

Ahora recordemos cómo funciona el grader:

La celda de calificación toma la definición de una función dentro del runtime de tu código (esto es,  busca una definición de función específica dentro de la sesión que tienes actualmente en tu notebook) y la envía al servidor para ser evaluada por el calificador automático específico.

Ahora, recuerda que él envia **solamente lo que está contenido dentro de la función**, es por esto que si importas una librería por fuera de la función el calificador automático no tiene forma de saber que esto ocurre.

Vamos a ver las formas correctas e incorrectas de importar las librerías para el calificador

<ins>**CORRECTA**</ins>

```
------------------------------------- comienzo de celda -----------------------------------------------------
def fit(X,y):
    from itertools import product
    def predict(X, t):
        return ...

    return ...
----------------------------------------fin de celda -----------------------------------------------------
```



**INCORRECTA**
```
------------------------------------- comienzo de celda -----------------------------------------------------
from itertools import product
def fit(X,y):
    def predict(X, t):
        return ...

    return ...
---------------------------------------fin de celda -----------------------------------------------------
```


**INCORRECTA**

```
------------------------------------- comienzo de celda -----------------------------------------------------

from itertools import product

--------------------------------------fin de celda -----------------------------------------------------

------------------------------------- comienzo de celda -----------------------------------------------------
def fit(X,y):
    def predict(X, t):
        return ...

    return ...
--------------------------------------fin de celda -----------------------------------------------------
```

Si te quedan dudas o sigue sin funcionarte me lo dices.

Hasta pronto!

**Reply by valenrami5** [reply]
Muchas gracias!! ya funciona :)
